Dispense med function

up2code3 opened the pull request:
## Description

_1) What does this PR do? 2) List out what you did._

Dispense medication 
1) finds medication
2) ensure med store matches user store
3)  ensures enough is in stock
4) Deducts quantity from stock
5) saves new quantity
6) creates a dispense log.

_Feature update/change/addition, bug fix, styling update, etc._

new feature

## Testing Instructions
end point and body
![Screenshot 2024-12-21 124740](https://github.com/user-attachments/assets/93ac7484-0743-4f6d-8c98-60b560386622)

note: I included , creating a dispenseLog model, and a "retrieve Dispense Log" controller , located directly under the dispense medication controller. I have not figured out how this works yet, I just new it would be the 2nd part of the dispense med function. Both are located in the medication controller.

up2code3 commented:
oooooooo.... I pulled up mongo compass, and checked the dispense logs there, it is in fact logging every time a medication was dispensed. So they only thing that is not working is my controller, to get the dispense log.

up2code3 commented:
![Screenshot 2024-12-21 131529](https://github.com/user-attachments/assets/bcceccd9-0728-4955-a7fe-fb4b38ee3c89)

GinaCastromonte commented:
<img width="635" alt="Screenshot 2024-12-23 at 12 28 19 PM" src="https://github.com/user-attachments/assets/ed5b6810-0d91-4bc5-b086-69422e4ddc45" />


Will the user be able to choose the date? If so, we need to add it to the frontend.  If not, then we don't need the dispenseDate on the model because you already have the timestamps. 

 Make sure dispense model has its own controller and route files to keep the code easier to maintain. When I was testing with Postman, the server kept crashing, and I found the the issue was due to the route order in the medication routes file. The route `/:id` was placed before other more specific routes like `/dispense-logs` and `/dispense`. Try to always place `/:id` at the end of the route definitions to avoid issues.

up2code3 commented:
ahh ok, Yea maybe just quantity on the dispense form for the front end.

up2code3 commented:
I just changed the dispense controller to use populate, and took store , lot , and ndc, off of the log model .

up2code3 commented:
Changed Create and Dispense **Controllers**, and 
Changed the Medication, and DispenseLog **Model.** 
got the opportunity to use populate in my controller, So i think i understand it a little better now. :)

GinaCastromonte commented:
Awesome job implementing all of this! 

- I kept getting an "invalid token" error, and I believe it might be related to the "expires in 1hr" setting in the token configuration. It might not be necessary in this case

- The issue with the dispense logs was that` /inventory/:id `was matching paths like dispense-logs because Express evaluates routes in the order they’re defined. By moving the dispense logs route above `/inventory/:id`, I makes sure it was processed first, which resolved the error.  OR just consider making a separate file.  

- Lastly, do we need to know which user dispensed the medication? If so, you might have to create another populate method to get user name, store, name of medication.  It depends on what  the frontend will need for reports (?)

up2code3 committed:
moved the dispense and getlog routes above all id routes FIXED

up2code3 committed:
created dispense controller and routes

up2code3 commented:
It is working and now it has its own controller and route files.


note for testing in post man , the route name changed, instead of
http://localhost:8000/api/v1/inventory/dispense-logs /dispense    (no more inventory in the name.)
it is now
http://localhost:8000/api/v1/dispense and /dispense-logs

up2code3 commented:
I kept getting an "invalid token" error, and I believe it might be related to the "expires in 1hr" setting in the token configuration. It might not be necessary in this case

I changed it to 24 hours,

up2code3 committed:
fixed get log to display more details for med and user

up2code3 commented:
I fix the dispense details for the log 
![Screenshot 2024-12-30 220211](https://github.com/user-attachments/assets/7593ae47-7db4-4adc-bcf7-25eea8a1f8c8)

up2code3 committed:
Merge branch 'main' into DispenseMedFunction

up2code3 commented:
conflicts resolved

up2code3 committed:
FixAllWorking

devandres-tech reviewed (approved):
changes look good
